cmd/zed/agents/zfs_retire.c

```diff
@@ -334,7 +334,7 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
 	 */
 	if (strcmp(class, "resource.fs.zfs.removed") == 0 ||
 	    (strcmp(class, "resource.fs.zfs.statechange") == 0 &&
-	    state == VDEV_STATE_REMOVED)) {
+	    (state == VDEV_STATE_REMOVED || state == VDEV_STATE_FAULTED))) {
```

Comment:
This doesn't seem quite right There are different states for REMOVED and FAULTED vdevs and the logic here is only for handling removed vdevs.  The code for handling faulted vdevs is at the end of this function.  We should figure out why it's occasionally not kicking in a spare.  Perhaps it's not in the FM_SUSPECT_FAULT_LIST or we're incorrectly hitting one of the `continue` cases in the `for` loop.

Reply:
So the "code below" appears to be reacting to "fault.io." events and triggering a state change to FAULTED (and also requesting a spare be deployed). Is zed (specifically the code below) the only source of a state change to FAULTED in zfs? If a device can become FAULTED via any other path (which appears to  be the case in the issue this PR is addressing) than there is no path to deploying a spare. The code above is responding to state changes. The proposed change above extends the range of state changes to respond to both REMOVED and FAULTED.

Reply:
I see what you're getting at. So there is a code path which should cover this case but depending on how you're testing I could see it potentially being unreliable. In addition to the zfs generated events the ZED consumes events from libudev which includes device removal events. These get converted to "resource.fs.zfs.removed" events here and are posted by the agent. That should cause this code path to be run and kick in a spare. Assuming you're getting the udev events.

Ideally I'd love to have zfs post these removal events itself instead of relying on udev. There is code in vdev_disk.c which tries to handle this but it effectively never runs since zfs_check_media_change() always returns zero. Unfortunately, I couldn't find a good kernel interface to use here would would distinguish between a normal I/O error and an error caused by a removed device.

Reply:
So it seems reasonable to me that we extend the logic here as proposed in this PR. If we reach this point in the zed logic and we have *already* transitioned a device to the FAULTED state, we should be deploying a spare if one is available. Once we are in the FAULTED state we are not going to receive anymore "fault.io" events, so the code lower down is not going to come into play and we will never deploy a spare.